### PR TITLE
Run nix language tests from upstream

### PR DESCRIPTION
**Pull request by layus**
At the moment, most tests fail, mainly due to missing builtins
It is nonetheless a good framework to start improving hnix.

### CODE REVIEW

**Comment by jwiegley** [comment]
This is fantastic, thank you!

**Comment by jwiegley** [comment]
There are a lot of other changes confounded with this pull request. I like the idea of integrating the Nix language tests, but let's do that as a separate thing. It's hard to review what you've done with these other changes in there as well.

**Comment by layus** [comment]
There is indeed a lot of related changes that may be split into different pull requests.

But it seemed to me that the individual commits where easy enough to review individually. Making the PR reviewable as a set of small commit-wise PRs.

I may split it later when I find the time to do so.

**Comment by layus** [comment]
Also, I was hoping for some feedback on the implementation before cleaning things up. Especially on NixLangTests.hs and Builtins.hs. I am quite new to Haskell.

**Comment by layus** [comment]
Voilà, this PR has been split and reduced to its minimal form.

**Comment by madjar** [comment]
Hi there, this looks pretty awesome!

I'm wondering (but that probably for some future refactoring) if it's possible to rework `NVFunction` so that `NVBuiltin1` doesn't have to be a separate case, and so that `NVBuiltin2` can be implemented with currying of `NVFunction` rather than as a separate case.

File: Nix/Builtins.hs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+-- | Evaluate a nix expression in the default context
+evalTopLevelExpr :: MonadFix m => NExpr -> m (NValue m)
+evalTopLevelExpr val = evalExpr val baseEnv
+
+baseEnv :: MonadFix m => ValueSet m
+baseEnv = Map.fromList $ [ ("builtins", Fix $ NVSet builtins) ] ++ topLevelBuiltins 
+  where
+    topLevelBuiltins = map mapping $ filter isTopLevel builtinsList
+    -- builtins = Map.fromList $ map mapping $ builtinsList
```
Comment (madjar) [inline Nix/Builtins.hs]: Some leftovers comments here?

File: Nix/Builtins.hs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+data BuiltinType = Normal | TopLevel
+data Builtin m = Builtin {kind :: BuiltinType, mapping :: (Text, NValue m) }
+
+isTopLevel :: Builtin m -> Bool
+isTopLevel b = case kind b of 
+                    Normal -> False
+                    TopLevel -> True
+
+--------------
```
Comment (madjar) [inline Nix/Builtins.hs]: It looks like some of this is implemented elsewhere. Can you remove what's been moved, and maybe add a comment explaining why the rest is there?

Comment (layus) [reply Nix/Builtins.hs]: It is implemented right below :-)

File: Nix/Builtins.hs (hunk 1 above)
Comment (layus) [reply Nix/Builtins.hs]: yes, again implemented right below :-)

**Comment by layus** [comment]
@madjar There are ways to avoid using different builtins, for example see https://stackoverflow.com/questions/16589145/typeclass-for-functions-with-different-numbers-of-arguments and http://okmij.org/ftp/Haskell/polyvariadic.html#polyvar-fn. This is a bit too far-fetched for me ATM.

There must also be a cleaner way to define the builtins without defining two function each time, but the right way to do this eludes me.

Anyway, thanks for reviewing !

**Comment by madjar** [comment]
Thinking out loud: considering we have
```hasAttr :: MonadFix m => NValue m -> NValue m -> m (NValue m)```

Could we define `prim_hasAttr` as:

```
builtin2 name function = Fix $ NVBuiltin1 name (\arg -> Fix $ NVBuiltin1 name (function arg))
prim_hasAttr = builtin2 "hasAttr" hasAttr
```

That way, calling `hasAttr` with one argument returns an `NVBuiltin1` that has the argument in its closure, rather than having it explicitly in a list. As a return, we only need `NVBuiltin` with no numbers :)

**Comment by layus** [comment]
@madjar Yes, I had the same idea but somehow I never managed to get the types right. It failed at invocation time in Eval.hs, for some obscure typing reasons that I could not debug or failed to understand.

I would be glad to learn from external improvements on this code

**Pull request merged**